src/extension/experimentService/experimentService.ts

```diff
@@ -0,0 +1,175 @@
+            .map(async (expConfig) => await this.executeExperiment(expConfig))
+        );
+
+        this.sendExperimentTelemetry(experimentResults);
+    }
+
+    public dispose() {
+        this.cancellationTokenSource.cancel();
+        this.cancellationTokenSource.dispose();
+    }
+
+    private constructor() {
+        this.endpointURL = "https://microsoft.github.io/vscode-react-native/experiments/experimentsConfig.json";
+        this.configName = "reactNativeToolsConfig";
+
+        this.promiseUtil = new PromiseUtil();
+        this.config = new Configstore(this.configName);
+        this.cancellationTokenSource = new vscode.CancellationTokenSource();
+        this.downloadedExperimentsConfig = null;
+
+        this.experimentsInstances = this.initializeExperimentsInstances();
+        this.downloadConfigRequest = this.retryDownloadExperimentsConfig();
+    }
+
+    private getAvailableExperiments(): Array<string> {
+        const cwd = path.join(__dirname, "experiments");
+        return glob.sync("*.js", { cwd })
+        .map(fullExpName => path.join(cwd, fullExpName));
+    }
+
+    private async executeExperiment(expConfig: ExperimentConfig): Promise<ExperimentResult> {
+        let curExperimentParameters = this.config.get(expConfig.experimentName);
+        let expInstance = this.experimentsInstances.get(expConfig.experimentName);
+
+        let expResult: ExperimentResult;
+        if (expInstance && expConfig.enabled) {
+            try {
+                expResult = await expInstance.run(expConfig, curExperimentParameters);
+                this.config.set(expConfig.experimentName, expResult.updatedExperimentParameters);
+            } catch (err) {
+                expResult = {
+                    resultStatus: ExperimentStatuses.FAILED,
+                    updatedExperimentParameters: expConfig,
+                };
+            }
+        } else {
+            expResult = {
+                resultStatus: ExperimentStatuses.DISABLED,
+                updatedExperimentParameters: expConfig,
+            };
+        }
+
+        return expResult;
+    }
+
+    private async retryDownloadExperimentsConfig(retryCount = 60): Promise<ExperimentConfig[]> {
+        try {
+            return await this.downloadExperimentsConfig();
+        } catch (err) {
+            if (retryCount < 1 || this.cancellationTokenSource.token.isCancellationRequested) {
+                throw err;
+            }
+
+            await this.promiseUtil.delay(2000);
+            return await this.retryDownloadExperimentsConfig(--retryCount);
+        }
+    }
+
+    private initializeExperimentsInstances(): Map<string, IExperiment> {
+        let expInstances = new Map<string, IExperiment>();
+        const availableExperiments = this.getAvailableExperiments();
+
+        availableExperiments.forEach(expPath => {
+            let expClass = require(expPath);
```

Comment:
Does this work when bundling?

Reply:
We implemented a [custom require function](https://github.com/microsoft/vscode-react-native/blob/preview/src/common/customRequire.ts) in the bundled extension. I asked Yuri about this place and he said that we could replace `require` with the `customRequire` function to be able to get modules dynamically. I think we will adapt this part of the code during adding of the experiment service to the Preview version.

Reply:
would webpack's ["magic comments"](https://webpack.js.org/api/module-methods/#magic-comments) work for this instead of custom require function?